### ISSUE TITLE
inconsistent computation of group/alt/opt box size given left or right arrows

### ISSUE THREAD

**Issue by jhoehle** [plantuml/plantuml#1097]
The computed size of alt/opt boxes differs on whether left or right arrows are used, as seen from the following examples.
```puml
@startuml
!pragma teoz true
skinparam ParticipantPadding 18
skinparam BoxPadding 20
box
participant A
end box
  A -> B
& B -> C
opt#transparent #pink bad /'in teoz'/
?o-> B : Pre
& B <-o? : yolo
'& C -[hidden]> C /' enlarge opt box '/
end
opt#transparent expand?
[o-> B : Pre
& B <-o? : yolo
'& C -[hidden]> C /' enlarge opt box '/
'& C -[hidden]> B
end
opt#transparent #20DD0090 OK
  A o-> B
& B <-o? : yolo
'& C -[hidden]> C /' enlarge opt box '/
end
opt#transparent #pink bad?
  A o-> B
& B <-o] : yolo
'& C -[hidden]> C /' enlarge opt box '/
end
@enduml
```
![Server view](https://www.plantuml.com/plantuml/png/fP8nJyCm48Nt-nKFHMgcgAOJ84JQj1i3sQiELsm5gubjEQuK-FLS8Y6G0ajOelZytlkUJjxsYMEwTYtTXCXDnqZ6Vo35gw7-ObtWo1rgqTXN6zYbchMshi7jtO_xoWzVyusYinyeV3km8-CqnYcmG_64YbQenfyz-P2o5DdrOZ2YpO9r5vnPOvrRDvN9rrJwqOK7rD5CtiV2bt9yzwsdV8KzYkERrTguauIA5SQr71i3oH_9a10fCSEPGHhgaeuJO6dqJ5Vz2iiswd1GwbxXvNbQnV-WWlXxVUKSS5e2sCf7NiSd)
Behaviour should be consistent. Summary of findings:

Teoz engine:
 - Short `?o-> B` is handled like the long arrow `[o->` causing _surprisingly large boxes_.

Standard engine:
 - The o (of o-> or ?o->) appears on the outside rim of the box.
 - The box grows more and more downwards for each hidden arrow
   (as if drawn transparently, instead of being there for effect only).

In both engines:
 - Uncommenting the hidden & arrow enlarges the box downwards.
   Explanation: because of the invisible U-turn self-referential arrow.
   => Better use the pattern `C-[hidden]>B` instead of `C-[hidden]>C` where possible!
 - The fourth case `<-o]` is not consistent with the second `[o->` one:
   The box enlarges to the left with `[o->` but not to the right with `<-o]`

There's a matter of visual design too: Perhaps long arrows should be ignored when computing the size of boxes? Would that allow nicer nesting?
In other words, perhaps the fourth case is actually OK and the second test case is wrongly expanded to the left?
This is not a new idea, forum user @The-Lum requested it in [QA-14106](https://forum.plantuml.net/14106/allow-long-out-coming-arrow-out-group-block-sequence-diagram?show=14106#q14106) .

I'd be in favour of the "ignore long arrows" visual effect.

**Comment by jimnelson372** [comment]
Hi @jhoehle,

The short arrow issue (the first pink group) with the latest release (V1.2024.5), which was a fix to the short arrow left side position fix needed to correct https://forum.plantuml.net/18832.   So your example will now look as follows:

![testeoz](https://github.com/plantuml/plantuml/assets/66284321/79f58a6f-0308-4107-97f7-60a404d68b92)

Why the <-] and <-o] isn't being completely included in that bottom pink group still needs to be looked into.

**Comment by jimnelson372** [comment]
I looked at at the problem with <-] and <-o] (which I'll call the "right edge" arrows since they extend to the end of the row) and at my level of understanding the layout, I'm not yet sure how to fix this problem.   

If you're curious about the implementation problem that is causing the right edge arrows to extend past their group (or, put another way, why their group isn't expanding to surround them), here is what I have found.

It is a circular issue:

1. The right edge of the diagram needs to know the width of the right edge arrows in order to know where the right side of the diagram will be.
2. Yet, the right edge arrows needs to know where the right edge of the diagram is in order to know how wide to draw.

Currently, it is working by calculating the diagram width (and the width of the group the arrow belongs in) as though the right edge arrows are simply short arrows. That is, as though their width is just the arrow head width with the text width and a little padding.  That's why the group's right edge is exactly the same as the first group's right edge in the teoz examples above.   Then, when it comes time to actually draw the right edge arrows, they are able to use that calculated diagram width to draw themselves all the way to the right edge.   The groups they belong to, unfortunately, do not know this, so they do not draw any wider.

In the above Teoz example, you can see that the right edge is slightly past the right edge of the C header, since the arrow text is short.   Change that last arrow from "yolo" to something much longer and you'll see the arrow itself determines the width of the diagram (in the first calculation), so the group and the arrow are both larger:

![testteozgroup_002](https://github.com/plantuml/plantuml/assets/66284321/cb316b34-fcff-433b-9929-94db2ab3ae58)

Fixing this is tricky, since I think if I try to tell the right edge arrow's group that it should be wider, that would then cause the diagram to calculate wider, which would then change the the width of the arrow and so on...

I'll give it more thought.  When I realized what was happening, I had to take a step back.

Jim N.

**Comment by arnaudroques** [comment]
Maybe we should add the notions of leftEdge() and getRightEdge() to the underlying real package:

```java
public interface Real {

    //...
    public Real getLeftEdge();
    public Real getRightEdge();
}
```

(Note that there are some commented-out getMaxAbsolute() and getMinAbsolute() methods there; I'm not sure why, but they seem related.)

The instances returned by `getLeftEdge()` and `getRightEdge()` will be special in the sense that you won't be able to call `addFixed()`, `addAtLeast()`, or `ensureBiggerThan()` on them.

However, they may be useful for denoting the positions of `[->` and `<-]` arrows.

What do you think about this approach? If you'd like, I can add these methods to the `Real` interface so you can check if they help.

**Comment by jimnelson372** [comment]
@arnaudroques, thanks, I can see how they could help.  Let me look at it once again today before seeing if the changes to Real will be needed (or preferred).   My last example at the end of my comment yesterday, which I discovered after I had written the earlier part of the comment, has made me realize (today) that I was wrong about the needed circular calculation I wrote about.   Clearly the positioning of the timeline for C turns out not to depend on the length of that right edge message at all -- nor on the short messages even if their text is made to be very long too.   

The calculation of the length of the right edge message, then, only seems to impact the group width.  So, I just want to double check whether C's right edge (in the example) is available to at <-] message when its width is being calculated, and if I can add a constraint to that arrow to ensureBiggerThan() that edge.   If that information is not available at the point where it is calculating its width, then yes, having a `getRightEdge()` would be helpful.   With that getRightEdge() function, I would then need to modify `getMaxX(..)` with: `point1.addFixed(width).ensureBiggerThan(point1.getRightEdge())`?   Or possibly there would be a method in `RealUtils` similar to `createOrigin()` that would be `createRightEdge()`? 

I think I've mentioned in another discussion where you've mentioned the constraints that I'm still not fully understanding its implementation.   Are you Real classes using a pattern that I can read about somewhere?  If so, please point me to it.

Regards,

Jim N.

**Comment by arnaudroques** [comment]
> I think I've mentioned in another discussion where you've mentioned the constraints that I'm still not fully understanding its implementation. Are you Real classes using a pattern that I can read about somewhere? If so, please point me to it.

No, the design and implementation are my own work. It might indeed be time for me to document this package "net.sourceforge.plantuml.real." Let me know if that would help you understand it better.

To summarize briefly, this package is independent of the rest of the PlantUML code and simply encodes the right side of the real number line. The origin point (O) has an abscissa of 0, and all other points (`Real`) are calculated from this origin. We merely define distance constraints between points.

Once all constraints are defined, we call the `compileNow()` method to calculate the actual values of the points/abscissas. This is done in a very basic way: we "push" the points to the right (thus increasing their values) until all constraints are satisfied. It's a very basic and straightforward approach, but caution is needed as it can quickly lead to cycles.



> I would then need to modify `getMaxX(..)` with: `point1.addFixed(width).ensureBiggerThan(point1.getRightEdge())`? Or possibly there would be a method in `RealUtils` similar to `createOrigin()` that would be `createRightEdge()`?


With the current approach, `point1.addFixed(width).ensureBiggerThan(point1.getRightEdge())` doesn't really make sense.

The reason is that `getRightEdge()` would return the actual right edge of *all* points, not just `point1`. Consequently, both `point1.getRightEdge()` and `pointxxx.getRightEdge()` would return the same value, representing the maximum extent of *all* points created.

It's admittedly not very logical to have such a method (`getRightEdge()`) on the `Real` class interface, but I'm unsure where else it would fit within the existing structure.

Therefore, you cannot ensure a point is bigger than the right edge because the right edge is the maximum value of all points created so far. Any point can only be positioned up to the current right edge, not beyond it.

I think I really should write a proper Javadoc about this :-)

**Comment by jimnelson372** [comment]
Oh, right.  It wouldn't need to use `ensureBiggerThan` since the goal is for the right side of the arrow to be at that `rightEdge`.

I haven't had a chance to look at the code again, but in the absence of having that `getRightEdge()` method, the `getMaxX()` that needs to be modified will just need to be able to find the far right `LivingSpace` and use its `posD`, which may be what the `getRightEdge()` would do without the calling component needing to have a reference to the list of living spaces.  I think I can solve it, but will let you know if I need that addition to Real.

I look forward to taking another look at the `Real.compileNow()` function to see how all the constraints are resolved, with the process you mentioned.

Regards,
Jim N

**Comment by jimnelson372** [comment]
I finally had a chance to return to this issue again today.   I wanted to give a brief update on what I'm finding.

It turned out to be very simple for the right edge arrow (in the `CommunicationExoTile`) to find the left most LivingSpace (via its `tileArguments`) to get the right side of it.  Here is what the original script for this issue becomes with the change I mentioned doing yesterday, along with an adjustment to the left left side as well.   Here is the original:

![image](https://github.com/plantuml/plantuml/assets/66284321/552474f1-977b-4a1f-8899-49f7c574aa42)

And here is what it looks like with my change:

![testteozgroup](https://github.com/plantuml/plantuml/assets/66284321/f3997a25-baa5-47fb-ac16-566b6a651596)

Unfortunately, as has been the case with some of these other grouping adjustments, this solution turns out to be incomplete, with some inconsistencies.  I don't have time to demonstrate all the problems this leaves, but here are a few:

- If C is put inside a `Box` as A is, it reveals that my right edge is not far enough right.   It should be symmetrical with the left edge padding to the left of A's Box, but it isn't _yet_.
- If there are other right edges with long text that pushes them further right, those right edge arrows that are not as long don't expand.  This would be fine, except when there are longer left side arrows, the left edge arrows do currently expand further left to match the start of those arrows.  So, either the left edge arrows should be changed, or the right edge should be.   Probably the right edge will needs to expand to the farthest right edge arrow.  However, that would bring me back to the problem I'm trying to solve with how to get the group to enclose that arrow as well.  I'm looking into a simpler approach of possibly keeping the current edge arrow logic close to what it now is, but having groups that contain these arrows also expand to the edges when drawn.

Jim N.

**Comment by jimnelson372** [comment]
> Maybe only the text of the arrow should be enclosed in the group, but not the right extremity of the arrow itself? 

If I'm understanding you correctly, that would leave it with what it is doing today.  Yet having the arrow extend past the end of the group is perceived as incorrect, since for left side groups the arrow is surrounded (though the arrow lands right on the left edge of the group in the current version, which is also seen as incorrect).   

My own preference is to have the group surround the arrow on both sides, but I don't tend to use this feature in my own UML diagrams, so it isn't a strong preference.  :)

I'd like to hear what @jhoehle and @The-Lum think can confirm that the current diagram (see my first comment above, or run the script online) is still not what it should be doing.

FYI: I haven't looked at the issue related to hidden arrows, that is also part of this issue.

Jim N.

**Comment by jimnelson372** [comment]
Here is a the other solution I mentioned wanting to try.   Just as the left edge [-> and right edge <-] messages are drawn meeting up with the final border positions, I did a simple test of ONLY change the group to draw itself to the same levels.  That is starting at Border1 if it contains a [-> message and ending at Border2 if it contains a <-].   I didn't add any margin in addition to that, so the result of the original example turns out to be:

![testteozgroup](https://github.com/plantuml/plantuml/assets/66284321/c884fc3e-2415-4504-8fcc-eb826e24d559)

This solution has the virtue of not drawing past the two border sides.   If you add a division, it's easier to see.  Here I just added a single divisor after the 2nd group that spans from the left edge to the right edge:


![testteozgroup_007](https://github.com/plantuml/plantuml/assets/66284321/1094fcac-d514-42e3-b371-865ebdf13830)

I think this looks good.  I can add margin to the group edge, but this at least surrounds the full arrow and does reveal that the edge arrows go all the way to the edge.

Let me know what you think.   There are still some symmetry problems if I add a box around B and C, the BoxPadding and ParticipantPadding do not seem to extend on the Right side.   So, I'm looking at that, which I hope is not too much more effort.  I started on this issue thinking it was one of the simpler problems, hoping to finish more of the teoz issues.

Jim N

**Comment by arnaudroques** [comment]
> I think this looks good.

I agree! But I'll let others give their opinion.
Thanks for your effort!

**Comment by jimnelson372** [comment]
One more example to demonstrate how this solution that merely extends the group edge (when it contains an edge arrow) works when there are longer text messages on both sides.   I show the same arrows with and without a group around them.  And you can see the differences between where the "o" side of the short arrows end (and are padded by their group) and where the edge arrows end.  And, the group extends to the edge on the right edge arrow in the "right edge arrows" division, while if you run that in the current release version, you'll see it doesn't.  I'll just show the improved version here.

![testteozgroup_003](https://github.com/plantuml/plantuml/assets/66284321/6a3f1b78-ae33-4522-ada9-74adc756f28a)

Here is the script:
```

@startuml
!pragma teoz true
skinparam ParticipantPadding 18
skinparam BoxPadding 20

participant A
participant B
participant C

== two short arrows ==
opt#transparent OK /'in teoz'/
?o-> B : Pre
& B <-o? : yolo
end
?o-> B : Pre
& B <-o? : yolo
opt#transparent OK /'in teoz'/
?o-> A : Pre
& A <-o? : yolo
end
?o-> A : Pre
& A <-o? : yolo
opt#transparent OK /'in teoz'/
?o-> A : longer text short
end
?o-> A : longer text short
== left edge arrows ==
opt#transparent OK
[o-> B : Pre
& B <-o? : yolo
end
[o-> B : Pre
& B <-o? : yolo
== right edge arrows ==
opt#transparent OK
  A o-> B
& B <-o] : yolo
end
  A o-> B
& B <-o] : yolo
== extra long text on right edge arrows ==
opt#transparent OK
  A o-> B
& B <-o] : This text extends the arrow length\n right edge arrow
end
  A o-> B
& B <-o] : This text extends the arrow length\n right edge arrow
== extra long text on right short arrows ==
opt#transparent OK /'in teoz'/
?o-> B : Pre
& B <-o? : This text extends the arrow length\n short arrow
end
?o-> B : Pre
& B <-o? : This text extends the arrow length\nshort arrow
@enduml
```

Regards,

Jim N.

**Comment by The-Lum** [comment]
Hi @jimnelson372 

It's a good compromise, even if initially I would have preferred that the arrows, coming from far away, were not included in the groups...
But your proposal is more consistent and manage all the arrow's cases.

Regards,
Th.

**Comment by jimnelson372** [comment]
Hello @arnaudroques, @The-Lum and @jhoehle,

I just realized I had not seen @The-Lum's Forum posts on this topic about groups surrounding edge arrows that @jhoehle linked to in the issue information above: [QA-14106](https://forum.plantuml.net/14106/).   

One of the examples in those discussions stood out to me, making me accept that my solution has problems.  That's the example from [QA-3443](https://forum.plantuml.net/3443/) with the following script:

```
loop
loop
Bob->Alice : hello
opt
[->Alice : external event
end opt
end loop
end loop
```

This currently generates this (in Puma):

![testteozgroup_010](https://github.com/plantuml/plantuml/assets/66284321/c9eb1fe8-1832-4a52-8d26-9a229976f5ee)

But it does draw a little better in the current release with Teoz:

![testteozgroup_010](https://github.com/plantuml/plantuml/assets/66284321/3fbcba00-2076-483f-b33d-0f276b57779f)

And when changed to [o->Alice it generates:

![testteozgroup_010](https://github.com/plantuml/plantuml/assets/66284321/cddec097-2f6a-4951-bf1a-ed859be45679)

While it's good that the edge arrow doesn't cause the group to expand, it's not clear to me how the size of the Opt group was determined, since even the **text of the edge arrow is outside the group.**   Normally, to contain that text, the group would have to expand, which we don't want.   We see better behavior on the short arrows, where the text remains closer to the participant and the group always surrounds it, even if long:

![testteozgroup_011](https://github.com/plantuml/plantuml/assets/66284321/47db2d96-2bbd-420e-b208-5a9805cf8885)

With that all demonstrated, this example convinces me that I should withdraw my proposed change, as it would have caused a regression on this subgroup behavior:

![testteozgroup_010](https://github.com/plantuml/plantuml/assets/66284321/51d02085-07c3-46dc-939b-47539a80bf89)

So, I'm actually now leaning towards **the solution @arnaudroques proposed several days ago** of not sizing groups according to their edges, but just their text; however it would be different from the current version implementation of the positioning of the left edge arrow text.   That text should remain within the group, so must stay on the right side of the arrow rather than the default left alignment, while the end of the arrow will continue farther out of the group towards the edge.   That happens easily today for the right edge arrows, since their text is left aligned (closer to the participant, still inside the group).  

I haven't implemented a test yet, but what I think will work is for groups to size as through the arrow is a short arrow AND for the text to draw as though it is also a short arrow (so the left edge text is closer to the participant life line), but to extend the line all the way to the edge.   That would make the left side work more like the right side.  To imagine this, consider the following diagram, where the text of the left edge message is outside its containing group, while the right edge text is inside it's containing group:

![testteozgroup_012](https://github.com/plantuml/plantuml/assets/66284321/9b584c44-6e9d-4bc0-8cfa-4a1255dd3f01)

Now look how the first Opt block changes when I make it a short arrow -- but then imagine that while the group and text remain just like this, the end of that arrow extends all the way to the edge.

![testteozgroup_012](https://github.com/plantuml/plantuml/assets/66284321/acc9d18a-fd2d-4c00-8798-ade98726b1d4)

Again, imagine that arrow in the first group extends all the way to the left edge (which I didn't draw).

I'm glad I delayed creating a pull request on this one.   I look forward for more feedback.

Best regards,

Jim Nelson
@jimnelson372

**Comment by arnaudroques** [comment]
> So, I'm actually now leaning towards **the solution @arnaudroques proposed several days ago** of not sizing groups according to their edges, but just their text; 

Yes, it's a nice solution and I have not changed my mind yet :-)

> however it would be different from the current version implementation of the positioning of the left edge arrow text. That text should remain within the group, so must stay on the right side of the arrow rather than the default left alignment, while the end of the arrow will continue farther out of the group towards the edge. That happens easily today for the right edge arrows, since their text is left aligned (closer to the participant, still inside the group).

Exactly. From an end-user perspective, including the arrow text within the group makes it clear that the arrow/message is part of the group. The fact that the arrow starts at the left edge of the diagram will not negatively impact the layout, even if the group is nested within another group.

It's challenging to be 100% certain that this is the best solution until we see the results with your changes. So good luck with your coding!

Congratulations on understanding PlantUML code, and thanks again for your contributions!

**Comment by jimnelson372** [comment]
Just and FYI.  I haven't forgotten about this issue.   I have coded some changes to position the text of edge arrows in the same area as the short arrow draws its text, but letting the arrow line and end extend to the edge, as we discussed.   I did a quick check with **pdiff** (to test against existing diagrams) and discovered this change will require a change to anchor positioning if one of the anchors is an edge arrow.  Here is the diagram that demonstrated the problem:

```
@startuml
!pragma teoz true

{start} [-> Bob : start doing things during duration
Bob -> Max : something
Max -> Bob : something else
{end} [-> Alice : finish

{start} <-> {end} : some time

@enduml
```

Notice that the end anchor is the edge arrow `[-> Alice : finish`.   It currently draws fine, when the "finish" is all the way to the left:

![image](https://github.com/plantuml/plantuml/assets/66284321/0280723d-efe1-4da2-9d94-591b97da98ee)

But, when I modified the text to draw closer to the arrow head, the anchor is positioned incorrectly:

![image](https://github.com/plantuml/plantuml/assets/66284321/365c6e7d-70b7-4a9e-938e-c1742fc12458)

Where it positioned, given the change in the position of "finish" is surprising to me.   I only saw this effect just last night, so I haven't had a chance to investigate it further.   It's a good example of the benefits of the new **pdiff** tool for developers, though.  :)  Thanks @arnaudroques!

To give a sample of the benefits of my fix, where groups are involved, see the difference in where the text falls in this diagram:

![image](https://github.com/plantuml/plantuml/assets/66284321/6758a4e6-a97a-4a15-a64b-3ba92c434583)

versus this diagram (with my change):

![image](https://github.com/plantuml/plantuml/assets/66284321/79cc3635-fbd7-4831-bcd6-15106bba0602)

Now the text "external event" is within the group box.   Once I have the anchors working again, I can look at how all the alignments of the arrows look with this new positioning.   Unfortunately, I will not be able to look at this again until next week (I believe).

Regards,

Jim N

**Comment by jimnelson372** [comment]
Hello @arnaudroques, @The-Lum and @jhoehle,

I've not been able to spend further time on any issues the past few weeks and I may have little time this week too, but I did have time to try a fix the the anchor problem I showed in my previous comment.   I have a simple fix that positions the anchor exactly where it was positioned previously, which will at least not be too much a a problem, though ultimately I think the logic of how anchors and their text should be positioned will need to be looked at further (maybe as a separate issue from the edge arrows).   In any case, let me show the results of the change.   Recall that I had found a diagram in *pdiff* that currently generates:

![image](https://github.com/plantuml/plantuml/assets/66284321/6fe9001a-8faf-46d1-a790-340b2bab0671)

and I showed that with my change to the edge arrows to make the text position just like with a short arrow, the anchor was drown in a bad spot:

![image](https://github.com/plantuml/plantuml/assets/66284321/025f6142-b73b-49a7-959a-e94481931f5c)

With a change I made last night, I have the anchor position itself exactly where it would have positioned itself originally, had I not made the text positioning change needed for the desired grouping of edge arrows.  

![image](https://github.com/plantuml/plantuml/assets/66284321/36f3a690-5ee0-42de-bb17-96b79dfb6483)

I think this will work for now, and in running pdiff on this new version, I don't see any new problems with my edge arrow solution.   I hope to have some time this evening to examine this again.

In the mean time, I realized I never showed the results of my current solution applied to the original script for this issue.   Here is how that diagram looked before, and here is how it would look with this change.  Again, the edge arrows would now position their message text exactly as though they were short arrows, but they'll extend to the edge.   The right side arrows don't change, since their message was already in that position, but the left side edge arrows now draw much better.

As reported above:

![image](https://github.com/plantuml/plantuml/assets/66284321/c97c2019-0d12-4c05-917b-5daba564471d)

And how it draws with my change:

![image](https://github.com/plantuml/plantuml/assets/66284321/82871c2f-5d94-44fa-aa7a-409b03962888)

I look forward to any feedback on this solution, including any suggestions for the anchor positioning (the logic of where the vertical arrow should be positioned, etc.)

Regards,

Jim N
@jimnelson372

**Comment by arnaudroques** [comment]
> I look forward to any feedback on this solution, including any suggestions for the anchor positioning (the logic of where the vertical arrow should be positioned, etc.)

Your solution looks fine, so you can go ahead!

**Comment by The-Lum** [comment]
Hi all,

_[This is an Issue Review]_ 👀 
This is now fixed on [V1.2024.6](https://github.com/plantuml/plantuml/releases/tag/v1.2024.6).

- Thanks @jhoehle for the report.
- Thanks @jimnelson372  for your work.

Regards.
